url/url.go

```diff
@@ -0,0 +1,189 @@
+/* package URL handles Helm-DM URLs
+
+Helm uses three kinds of URLs:
+
+	- Fully qualified (long) names: https://example.com/foo/bar-1.2.3.tgz
+	- Short names: helm:example.com/foo/bar#1.2.3
+	- Local names: file:///foo/bar
+
+This package provides utilities for working with this type of URL.
+*/
+package url
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+const (
+	SchemeHTTP  = "http"
+	SchemeHTTPS = "https"
+	SchemeHelm  = "helm"
+	SchemeFile  = "file"
+)
+
+// TarNameRegex parses the name component of a URI and breaks it into a name and version.
+//
+// This borrows liberally from github.com/Masterminds/semver.
+const TarNameRegex = `([0-9A-Za-z\-_/]+)-(v?([0-9]+)(\.[0-9]+)?(\.[0-9]+)?` +
+	`(-([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?` +
+	`(\+([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?)(.tgz)?`
+
+var tnregexp *regexp.Regexp
+
+func init() {
+	tnregexp = regexp.MustCompile("^" + TarNameRegex + "$")
+}
+
+type URL struct {
+	// The scheme of the URL. Typically one of http, https, helm, or file.
+	Scheme string
+	// The host information, if applicable.
+	Host string
+	// The bucket name
+	Bucket string
+	// The chart name
+	Name string
+	// The version or version range.
+	Version string
+
+	// If this is a local chart, the path to the chart.
+	LocalRef string
+
+	isLocal bool
+
+	original string
+}
+
+func Parse(path string) (*URL, error) {
+
+	// Check for absolute or relative path.
+	if path[0] == '.' || path[0] == '/' {
+		return &URL{
+			LocalRef: path,
+			isLocal:  true,
+			original: path,
+		}, nil
+	}
+
+	// TODO: Do we want to support file:///foo/bar.tgz?
```

Comment:
I'll take a look. There was some oddball case I was trying to avoid, but I can't remember what it was. It was something about it getting parsed as an opaque under one set of circumstances, and as a path on the other. But it deserves another look.